benchmark/Cajita_FastFourierTransformPerformance.cpp

```diff
@@ -0,0 +1,240 @@
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "Cabana_BenchmarkUtils.hpp"
+
+#include <Cajita.hpp>
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Random.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <mpi.h>
+
+using namespace Cajita;
+
+//---------------------------------------------------------------------------//
+// Performance test.
+template <class Device, class HostBackendType>
+void performanceTest( std::ostream& stream,
+                      std::vector<double> grid_sizes_per_dim, MPI_Comm comm,
+                      const std::string& test_prefix )
+{
+    // Domain size setup
+    std::array<double, 3> global_low_corner = { 0.0, 0.0, 0.0 };
+    std::array<double, 3> global_high_corner = { 1.0, 1.0, 1.0 };
+    std::array<bool, 3> is_dim_periodic = { true, true, true };
+
+    int num_grid_size_per_dim = grid_sizes_per_dim.size();
+
+    // number of runs in test loops
+    int num_runs = 10;
+
+    // get MPI rank
+    int comm_rank;
+    MPI_Comm_rank( comm, &comm_rank );
+
+    // create timers
+    Cabana::Benchmark::Timer setup_timer( test_prefix + "setup",
+                                          num_grid_size_per_dim );
+
+    Cabana::Benchmark::Timer transforms_timer( test_prefix + "transforms",
+                                               num_grid_size_per_dim );
+    // loop over the grid sizes
+    for ( int p = 0; p < num_grid_size_per_dim; ++p )
+    {
+        double cell_size = 1.0 / grid_sizes_per_dim[p];
+        auto global_mesh = createUniformGlobalMesh(
+            global_low_corner, global_high_corner, cell_size );
+
+        // Create the global grid
+        DimBlockPartitioner<3> partitioner;
+        auto global_grid =
+            createGlobalGrid( comm, global_mesh, is_dim_periodic, partitioner );
+
+        // Create a local grid
+        auto local_grid = createLocalGrid( global_grid, 0 );
+        auto owned_space = local_grid->indexSpace( Own(), Cell(), Local() );
+        auto ghosted_space = local_grid->indexSpace( Ghost(), Cell(), Local() );
+
+        auto vector_layout = createArrayLayout( local_grid, 2, Cell() );
+        auto lhs = createArray<double, Device>( "lhs", vector_layout );
+        auto lhs_view = lhs->view();
+        auto lhs_host =
+            createArray<double, typename decltype( lhs_view )::array_layout,
+                        Kokkos::HostSpace>( "lhs_host", vector_layout );
+        auto lhs_host_view = Kokkos::create_mirror_view( lhs_view );
+        uint64_t seed = global_grid->blockId() +
+                        ( 19383747 % ( global_grid->blockId() + 1 ) );
+        using rnd_type = Kokkos::Random_XorShift64_Pool<Kokkos::HostSpace>;
```

Comment:
Good point. I don't think there is any issue - switched to using device

Reply:
It's failing now because the loop below needs to be a Kokkos kernel and use the device views (shouldn't need any of the host views anymore). Also, pass the memory space from the device rather than the device itself.

Reply:
Ah, that's great we can remove the host view and copy. Fixed to use a Kokkos kernel on the device, and got rid of host view